tango/integrations/wandb/workspace.py

```diff
@@ -0,0 +1,429 @@
+        """
+        app_url = self.wandb_client.client.app_url
+        if app_url.endswith("/"):
+            app_url = app_url[:-1]
+        return f"{app_url}/{self.entity}/{self.project}"
+
+    def _get_unique_id(self, step_or_unique_id: Union[Step, str]) -> str:
+        if isinstance(step_or_unique_id, Step):
+            unique_id = step_or_unique_id.unique_id
+        else:
+            unique_id = step_or_unique_id
+        return unique_id
+
+    def step_dir(self, step_or_unique_id: Union[Step, str]) -> Path:
+        unique_id = self._get_unique_id(step_or_unique_id)
+        path = self.steps_dir / unique_id
+        path.mkdir(parents=True, exist_ok=True)
+        return path
+
+    def work_dir(self, step: Step) -> Path:
+        path = self.step_dir(step) / "work"
+        path.mkdir(parents=True, exist_ok=True)
+        return path
+
+    def step_info(self, step_or_unique_id: Union[Step, str]) -> StepInfo:
+        unique_id = self._get_unique_id(step_or_unique_id)
+        if unique_id in self._running_step_info:
+            return self._running_step_info[unique_id]
+        step_info = self._get_updated_step_info(
+            unique_id,
+            step_name=step_or_unique_id.name if isinstance(step_or_unique_id, Step) else None,
+        )
+        if step_info is None:
+            raise KeyError(step_or_unique_id)
+        else:
+            return step_info
+
+    def step_starting(self, step: Step) -> None:
+        if wandb.run is not None:
+            raise RuntimeError(
+                "There is already a W&B run initialized, cannot initialize another one."
+            )
+
+        work_dir = self.work_dir(step)
+
+        lock_path = self.step_dir(step) / "lock"
+        lock = FileLock(lock_path, read_only_ok=True)
+        lock.acquire_with_updates(desc=f"acquiring lock for '{step.name}'")
+        self.locks[step] = lock
+
+        step_info = self._get_updated_step_info(step.unique_id) or StepInfo.new_from_step(step)
+        if step_info.state not in {StepState.INCOMPLETE, StepState.FAILED, StepState.UNCACHEABLE}:
+            raise RuntimeError(
+                f"Step '{step.name}' is trying to start, but it is already {step_info.state}. "
+                "If you are certain the step is not running somewhere else, delete the lock "
+                f"file at {lock_path}."
+            )
+
+        try:
+            # Initialize W&B run for the step.
+            wandb.init(
+                name=step_info.step_name,
+                job_type=RunKind.STEP.value,
+                group=step.unique_id,
+                dir=str(work_dir),
+                entity=self.entity,
+                project=self.project,
+                # For cacheable steps we can just use the step's unique ID as the W&B run ID,
+                # but not for uncacheable steps since those might be ran more than once, and
+                # and will need a unique W&B run ID each time.
```

Comment:
Do uncacheable steps need a wandb run ID at all? They are usually just folded into the step that has them as a dependency ... except for uncacheable leaf steps.

Reply:
Not necessarily, but I think it's better this way as it results in a complete execution graph in W&B.